lib/ansible/modules/pip.py

```diff
@@ -611,6 +613,61 @@ def setup_virtualenv(module, env, chdir, out, err):
     return out, err, cmd
 
 
+def _normalize_vcs_packages(module: AnsibleModule,
+                            package_list: list[Package],
+                            pip: list[str]) -> list[Package]:
+    """Return list[Package] where all vcs URL packages are converted to actual package names"""
+    vcs_packages = [str(pkg) for pkg in package_list if _is_vcs_url(str(pkg))]
+
+    if not vcs_packages:
+        return package_list
+
+    other_packages = [pkg for pkg in package_list if not _is_vcs_url(str(pkg))]
```

Comment:
Would the `__contains__()` check be more efficient than using regexes?
```suggestion
    other_packages = [
        pkg for pkg in package_list
        if str(pkg) not in vcs_packages
    ]
```

Reply:
Also, this is still computed early, before we know we'd take the successful path. So it's best to move it to the end.

Reply:
Moved to the end with the other generator changes o7